File: lib/JFFNetwork/Errors/JHttpError.m
#import "JHttpError.h"

@implementation JHttpError

- (instancetype)init
{
    [self doesNotRecognizeSelector:_cmd];
    return nil;
}

- (instancetype)initWithDescription:(NSString *)description
                             domain:(NSString *)domain
                               code:(NSInteger)code
{
    [self doesNotRecognizeSelector:_cmd];
    return nil;
}

- (instancetype)initWithDescription:(NSString *)description
                               code:(NSInteger)code
{
    return [super initWithDescription:description
                               domain:@"com.just_for_fun.library.http"
                                 code:code];
}

- (instancetype)initWithHttpCode:(CFIndex)statusCode
{
    return [self initWithDescription:NSLocalizedString(@"JFF_HTTP_ERROR", nil)
                                code:statusCode];
}

- (instancetype)copyWithZone:(NSZone *)zone
{
    JHttpError *copy = [super copyWithZone:zone];
    
    if (copy) {
        copy->_context = [_context copyWithZone:zone];
    }
    
    return copy;
}

- (NSString *)errorLogDescription
{
    return [[NSString alloc] initWithFormat:@"%@ Http code:%d cantext:%@",
            [self localizedDescription],
            self.code,
            _context];
}

- (BOOL)isHttpNotChangedError
{
    return (self.code == 304);
}

- (BOOL)isServiceUnavailableError
{
    return (self.code == 503);
}

- (BOOL)isInternalServerError
{
    return (self.code == 500);
}

@end

File: iAsync.Swift/lib/JNetwork/JNetwork/NSError/JHttpError.swift
//
//  JHttpError.swift
//  Wishdates
//
//  Created by Vladimir Gorbenko on 18.08.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

import JUtils

public class JHttpError : JNetworkError {
    
    private let context: AnyObject
    private let httpCode: CFIndex
    
    @objc public required init(httpCode: CFIndex, context: AnyObject) {
        
        self.httpCode = httpCode
        self.context  = context
        
        super.init(
            description: "J_HTTP_ERROR",
            domain     : "com.just_for_fun.library.http",
            code       : httpCode)
    }

    required public init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func copyWithZone(zone: NSZone) -> AnyObject {
        
        return self.dynamicType(httpCode: httpCode, context: context)
    }
    
    func isHttpNotChangedError() -> Bool {
        
        return code == 304
    }
    
    func isServiceUnavailableError() -> Bool {
        
        return code == 503
    }
    
    func isInternalServerError() -> Bool {
        
        return code == 500
    }
    
    func isNotFoundError() -> Bool {
        
        return code == 404
    }
    
    override public var errorLogDescription: String {
        
        return "\(self.dynamicType) : \(localizedDescription) Http code:\(code) context:\(context.description)"
    }
}
